include/lsst/afw/math/statistics/Framework.h

```diff
@@ -0,0 +1,615 @@
+                                  WeightPixelT wgt) {
+        std::get<0>(t).accumulateClipped(img, msk, var, wgt);
+    }
+};
+
+/**
+ * Call `accumulateClipped` member function for every member of `ProductSet` argument `t`.
+ */
+template <typename... Args, typename ImagePixelT, typename MaskPixelT, typename VariancePixelT,
+          typename WeightPixelT>
+void accumulateClippedHelper(ProductSet<Args...>& t, ImagePixelT img, MaskPixelT msk, VariancePixelT var,
+                             WeightPixelT wgt) {
+    AccumulateClippedHelper<decltype(t), sizeof...(Args)>::accumulateClipped(t, img, msk, var, wgt);
+}
+
+/**
+ * Helper class to call `combine` member function for every member of a `ProductSet`.
+ */
+template <typename T, size_t N>
+struct CombineHelper {
+    template <typename Product>
+    static void combine(T& t, Product const& other) {
+        CombineHelper<T, N - 1>::combine(t, other);
+        std::get<N - 1>(t).combine(other);
+    }
+};
+
+template <typename T>
+struct CombineHelper<T, 1> {
+    template <typename Product>
+    static void combine(T& t, Product const& other) {
+        std::get<0>(t).combine(other);
+    }
+};
+
+/**
+ * Call `combine` member function for every member of `ProductSet` argument `t`.
+ */
+template <typename... Args, typename Product>
+void combineHelper(ProductSet<Args...>& t, Product const& other) {
+    CombineHelper<decltype(t), sizeof...(Args)>::combine(t, other);
+}
+
+/**
+ * Base class for statistics products (e.g. Sum, Count, Variance, etc.).
+ *
+ * A Product is templated on the intermediate Products it requires (or None),
+ * and defines an `accumulate` member function (default noop).
+ */
+template <typename... Ts>
+class Product {
+public:
+    using Intermediate = typename FlattenOne<ProductSet<Ts...>>::type;
+
+    /**
+     * Combine two products
+     */
+    template <typename T>
+    void combine(T const&) { /* noop for abstract products */
+    }
+
+    /**
+     * Process valid value (pixel, mask, variance and/or weight)
+     */
+    template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT, typename WeightPixelT>
+    void accumulate(ImagePixelT, MaskPixelT, VariancePixelT, WeightPixelT) { /* noop for abstract products */
+    }
+
+    /**
+     * Process clipped value (pixel, mask, variance and/or weight)
+     */
+    template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT, typename WeightPixelT>
+    void accumulateClipped(ImagePixelT, MaskPixelT, VariancePixelT,
+                           WeightPixelT) { /* noop for abstract products */
+    }
+};
+
+/**
+ * Class that represents a concrete Product.
+ *
+ * Creates an optimal packing of unique intermediate products and provides
+ * an `accumulate` member function that accumulates each one.
+ *
+ * Results may be extracted from such objects using the non-member `extract` function.
+ */
```

Comment:
Please explain the difference between `ConcreteProduct` and `Product`. When might you subclass one vs. the other?